keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2861 @@
+
+###### Alternative: `PolicyConfiguration` kind with config reference
+
+Require two resources:
+
+- A `PolicyConfiguration` resources contains the match criteria and a reference
+  to a configuration resource.
+- A configuration resource that contains only configuration (no match criteria).
+
+Example:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: PolicyConfiguration
+metadata:
+  name: "xyzlimit-scale.example.com"
+  labels:
+    policyConfig: xyzlimit
+spec:
+  match:
+    resourceRules:
+      - apiGroups: ["*"]
+        apiVersions: ["*"]
+        resources: ["*/scale"]
+  config:
+    name: xyzlimit-scale-config.example.com
+```
+
+```yaml
+apiVersion: rules.example.com/v1
+kind: XyzLimit
+metadata:
+  name: "xyzlimit-config.example.com"
+spec:
+  allowed: ["a", "b", "c"]
+  banned: ["x", "y", "z"]
+  xyz:
+    fuzzFactor: 0.8
+    reticulate: true
+```
+
+Note that for policy configurations that require no parameterization, only the
+`PolicyConfiguration` is needed.
+
+The `metadata.labels.policyConfig` label on `PolicyConfiguration` is needed so
+that each `ValidatingAdmissionPolicy` use a label selector to reference the
+policy configurations that configure it.
+
+Pros:
+
+- Matching criteria is fully defined and validated in a builtin type.
+- The configuration data that the policy definition is treated as an opaque
+  payload from an apiserver perspective, it passes it to the policy definition
+  so that CEL expressions can access it, but there is no code in the apiserver
+  that cares about the structure of the configuration CRD.
+- If later Kubernetes OpenAPIv3 supports `$ref`, there is a migration path from
+  this approach to the `$ref` approach (below).
+
+Cons:
+
+- Requires two resources for each configuration of a policy
+  - Cannot atomically change the match rules and configuration of a policy configuration
+- Must establish relationships between `ValidatingAdmissionPolicy` and
+  `PolicyConfiguration` resources.
+
+###### Alternative: OpenAPIv3 `$ref` in CRDs
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: replicalimit.rules.example.com
+spec:
+  group: example.com
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                config:
+                  type: object
+                    maxReplicas:
+                      type: int
+                match:
+                  $ref: "#/components/schemas/matchrules"
+  scope: Cluster
+  names:
+    kind: ReplicaLimit
+```
+
+Pros:
+
+- Match rule schema is owned by Kubernetes, so as it evolves, CRDs automatically
+  pick up changes.
+
+Cons:
+
+- CRDs to not yet support OpenAPIv3 `$ref`s, so support would need to be added,
+  presumably this would require a separate KEP.
+
+###### Alternative: `/matchRules` subresource
```

Comment:
It doesn't sound to me like this solves the problems with the first or third and instead it adds a new one?

(If we were doing this in-tree and we wanted a "no bindings" system this would be the way to go, I think. But neither are true?)

Reply:
Eliminating this alt too. I think everyone has agreed that it's not viable.